### PR TITLE
[FEATURE] Implement Text Editor for Application Notes

### PR DESCRIPTION
**Pull request by Natata08**
# [FEATURE] Implement Text Editor for Application Notes

---

## Description

This PR implements a rich text editor component for managing application notes using React-Quill. The component provides a user-friendly interface for creating, editing, and viewing notes associated with job applications.
Also, the EmptyState component was added to be used with all tabs which don't have content.

Problem: Need a way to maintain detailed notes for each job application with rich text formatting
Impact: Users can now create and maintain formatted notes with features like headers, lists, and text styling

---

### Frontend:
Notes Component:
- Implemented rich text editor using React-Quill
- Added edit/read modes with proper state management
- Implemented empty state display

Editor Styling:
- Added comprehensive theme-aware styling for the editor
- Implemented hover states and active states for toolbar buttons
- Ensured consistent styling with the application's theme

Empty State Component:
- Created a separate component for empty state display
- Consistent styling with other empty states in the application

---

### Empty state
screencapture-localhost-3000-applications-28-2024-11-24-09_54_54

screencapture-localhost-3000-applications-28-2024-11-24-09_54_54

### Editor in read mode
screencapture-localhost-3000-applications-39-2024-11-21-20_53_08

screencapture-localhost-3000-applications-39-2024-11-21-20_52_58

### Editor in edit mode with toolbar
screencapture-localhost-3000-applications-39-2024-11-21-20_52_30

screencapture-localhost-3000-applications-39-2024-11-21-20_52_47

### On mobile:
Mobile
Mobile
Mobile

---

## Testing Steps

1. Navigate to any application's details page
2. Click on the Notes tab
3. Verify empty state is displayed correctly
4. Test editor functionality:
 - Click Add button
 - Try different text formatting options
 - Save changes
 - Cancel changes
5. Test responsive behavior on different screen sizes
6. Verify theme consistency in both light and dark modes
---

## Issues Addressed

Closes [Trello ticket link for Notes feature](https://trello.com/c/CNimWft4)

---

## Additional Notes

The next step is creating endpoints to get, put, and update notes from the Database.

### CODE REVIEW

File: frontend/app/applications/[id]/notes/Notes.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        ['clean'],
+      ],
+    },
+    theme: 'snow',
+    readOnly: !isEditing,
+  })
+
+  useEffect(() => {
+    if (quill) {
+      quill.on('text-change', () => {
```
Comment (pankaj28843) [inline frontend/app/applications/[id]/notes/Notes.js:101]: Shouldn't you extract this event handler as a function and in useEffect's return register `quill.off` for this event and the same function??

File: frontend/app/applications/[id]/notes/Notes.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import EmptyNotes from './EmptyNotes'
+
+const Notes = ({ applicationId }) => {
+  const [value, setValue] = useState('')
+  const [lastSavedContent, setLastSavedContent] = useState(null)
+  const [isEditing, setIsEditing] = useState(false)
+  const [isSaving, setIsSaving] = useState(false)
+  const theme = useTheme()
+
+  const editorStyles = {
```
Comment (pankaj28843) [inline frontend/app/applications/[id]/notes/Notes.js:16]: See if useMemo on this gives better performance, now that you know how to do performance optimization for React components, it is a good idea to do profiling of such new components that rely on external JS libraries for rich features like text WYSIWYG editors.

File: frontend/app/applications/[id]/notes/Notes.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    }
+  }, [
+    quill,
+    theme.palette.mode,
+    isEditing,
+    theme.palette.text.primary,
+    theme.palette.divider,
+  ])
+
+  const handleSave = async () => {
```
Comment (pankaj28843) [inline frontend/app/applications/[id]/notes/Notes.js:135]: wrap it with `useCallback`

File: frontend/app/applications/[id]/notes/Notes.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    } finally {
+      setIsSaving(false)
+    }
+  }
+
+  const handleEdit = () => {
+    setIsEditing(true)
+  }
+
+  const handleCancel = () => {
```
Comment (pankaj28843) [inline frontend/app/applications/[id]/notes/Notes.js:152]: Same thing about `useCallback`. Optimize for re-render, as much as possible.

**Review by pankaj28843 — CHANGES_REQUESTED** [review]
I suggest doing some profiling on this, and let's merge only after that.

**Review by magdazelena — COMMENTED** [review]
I agree with @pankaj28843 comments, added a small thing of my own.

Great job though!

File: frontend/app/applications/[id]/notes/Notes.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+              Cancel
+            </Button>
+            <Button
+              variant="contained"
+              size="small"
+              onClick={handleSave}
+              disabled={isSaving}
+            >
+              {isSaving ? 'Saving...' : 'Save'}
+            </Button>
```
[the remark this answers is not in the document]
Comment (Natata08) [reply frontend/app/applications/[id]/notes/Notes.js:181]: I hope I fixed it correctly

File: frontend/app/applications/[id]/notes/Notes.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    theme.palette.mode,
+    isEditing,
+    theme.palette.text.primary,
+    theme.palette.divider,
+  ])
+
+  const handleSave = useCallback(async () => {
+    setIsSaving(true)
+    try {
+      console.log('notes:', value)
```
Comment (pankaj28843) [inline frontend/app/applications/[id]/notes/Notes.js:157]: Did you forget to clean up this?

**Review by pankaj28843 — DISMISSED** [review]
Looks good! 💯 💪🏻

**Review by myselfankitaa — DISMISSED** [review]

**Review by pankaj28843 — APPROVED** [review]

**Review by myselfankitaa — APPROVED** [review]

**Pull request merged**